Lib/dataclasses.py

```diff
@@ -357,19 +361,19 @@ def _create_fn(name, args, body, *, globals=None, locals=None,
     return locals[name]
 
 
-def _field_assign(frozen, name, value, self_name):
+def _field_assign(frozen, name, value, self_name, object_expression):
     # If we're a frozen class, then assign to our fields in __init__
     # via object.__setattr__.  Otherwise, just use a simple
     # assignment.
     #
     # self_name is what "self" is called in this function: don't
     # hard-code "self", since that might be a field name.
     if frozen:
-        return f'object.__setattr__({self_name},{name!r},{value})'
+        return f'{object_expression}.__setattr__({self_name},{name!r},{value})'
```

Comment:
I think the only change you need to make in this file is to use `__builtins__.object` instead of `object` here. Since identifiers that start with double underscores are reserved for Python, we don't need to support a field named `__builtins__`.

Reply:
That's indeed a right thing do, I completely forgot about `__builtins__`.

However, surprisingly (at least for me) you should do `__builtins__['object']__`, not `__builtins__.object` since `__builtins__` is a dictionary inside `exec`, not a module: “If the globals dictionary does not contain a value for the key `__builtins__`, a reference to the dictionary of the built-in module builtins is inserted under that key.” — https://docs.python.org/3/library/functions.html#exec